src/inspector_socket.cc

```diff
@@ -0,0 +1,762 @@
+#endif
+
+  // Freed in write_request_cleanup
+  uv_buf_t* buf = reinterpret_cast<uv_buf_t*>(malloc(sizeof(uv_buf_t)));
+  uv_write_t* req = reinterpret_cast<uv_write_t*>(malloc(sizeof(uv_write_t)));
+  CHECK_NE(buf, nullptr);
+  CHECK_NE(req, nullptr);
+  memset(req, 0, sizeof(*req));
+  buf->base = reinterpret_cast<char*>(malloc(len));
+
+  CHECK_NE(buf->base, nullptr);
+
+  memcpy(buf->base, msg, len);
+  buf->len = len;
+  req->data = buf;
+
+  uv_stream_t* stream = reinterpret_cast<uv_stream_t*>(&inspector->client);
+  return uv_write(req, stream, buf, 1, write_cb) < 0;
+}
+
+// Constants for hybi-10 frame format.
+
+typedef int OpCode;
+
+const OpCode kOpCodeContinuation = 0x0;
+const OpCode kOpCodeText = 0x1;
+const OpCode kOpCodeBinary = 0x2;
+const OpCode kOpCodeClose = 0x8;
+const OpCode kOpCodePing = 0x9;
+const OpCode kOpCodePong = 0xA;
+
+const unsigned char kFinalBit = 0x80;
+const unsigned char kReserved1Bit = 0x40;
+const unsigned char kReserved2Bit = 0x20;
+const unsigned char kReserved3Bit = 0x10;
+const unsigned char kOpCodeMask = 0xF;
+const unsigned char kMaskBit = 0x80;
+const unsigned char kPayloadLengthMask = 0x7F;
+
+const size_t kMaxSingleBytePayloadLength = 125;
+const size_t kTwoBytePayloadLengthField = 126;
+const size_t kEightBytePayloadLengthField = 127;
+const size_t kMaskingKeyWidthInBytes = 4;
+
+static void encode_frame_hybi17(const char* message,
+                                size_t data_length,
+                                int masking_key,
+                                bool compressed,
+                                char** output,
+                                size_t* output_len) {
+  std::vector<char> frame;
+  OpCode op_code = kOpCodeText;
+  int reserved1 = compressed ? kReserved1Bit : 0;
+  frame.push_back(kFinalBit | op_code | reserved1);
+  char mask_key_bit = masking_key != 0 ? kMaskBit : 0;
+  if (data_length <= kMaxSingleBytePayloadLength) {
+    frame.push_back(static_cast<char>(data_length) | mask_key_bit);
+  } else if (data_length <= 0xFFFF) {
+    frame.push_back(kTwoBytePayloadLengthField | mask_key_bit);
+    frame.push_back((data_length & 0xFF00) >> 8);
+    frame.push_back(data_length & 0xFF);
+  } else {
+    frame.push_back(kEightBytePayloadLengthField | mask_key_bit);
+    char extended_payload_length[8];
+    size_t remaining = data_length;
+    // Fill the length into extended_payload_length in the network byte order.
+    for (int i = 0; i < 8; ++i) {
+      extended_payload_length[7 - i] = remaining & 0xFF;
+      remaining >>= 8;
+    }
+    frame.insert(frame.end(), extended_payload_length,
+                 extended_payload_length + 8);
+    assert(!remaining);
+  }
+
+  if (masking_key != 0) {
+    const char* mask_bytes = reinterpret_cast<char*>(&masking_key);
+    frame.insert(frame.end(), mask_bytes, mask_bytes + 4);
+    for (size_t i = 0; i < data_length; ++i)  // Mask the payload.
+      frame.push_back(message[i] ^ mask_bytes[i % kMaskingKeyWidthInBytes]);
```

Comment:
So, I infer that endianness is not relevant but this code still looks kind of iffy.  If nothing else, replacing `kMaskingKeyWidthInBytes` with `sizeof(masking_key)` seems more appropriate; so does using `uint32_t` instead of `int`.